addon/src/-private/array.ts

```diff
@@ -161,11 +161,8 @@ class TrackedArray<T = unknown> {
 
         let index = convertToInt(prop);
 
-        if (index !== null) {
-          self.#dirtyStorageFor(index);
-          setValue(self.#collection, null);
```

Comment:
When reading a value from the array, the `self.#collection` storage is always read in addition to the index-specific storage.

Therefore, there is no benefit to invalidating the per-index storage immediately before invalidating the collection-wide storage.